### PR TITLE
Missing a `state` url parameter

### PR DESCRIPTION
**Pull request by slava-tre**
`state` url parametr - optional arbitrary string that is returned as a URL parameter in a redirect_uri

### CODE REVIEW

**Comment by khamidou** [comment]
Thank you for the pull request @slava-tre! It looks good to me, but the SDK is compiled from coffeescript, so you'd need to make the same changes to [nylas.coffee](https://github.com/nylas/nylas-nodejs/blob/master/nylas.coffee#L54)

File: lib/nylas.js [hunk 1]
```diff
@@ -102,7 +102,10 @@
       if (options.trial == null) {
         options.trial = false;
       }
-      return this.apiServer + "/oauth/authorize?client_id=" + this.appId + "&trial=" + options.trial + "&response_type=code&scope=email&login_hint=" + options.loginHint + "&redirect_uri=" + options.redirectURI;
+      if (options.state == null) {
+        options.state = '';
```
Comment (khamidou) [inline lib/nylas.js]: I'd prefer to not add `state` to the URL if it isn't defined.

**Comment by slava-tre** [comment]
Hi @khamidou! Is my updates - OK?

File: nylas.coffee [hunk 2]
```diff
@@ -51,6 +51,9 @@ class Nylas
     throw new Error("urlForAuthentication() requires options.redirectURI") unless options.redirectURI?
     options.loginHint ?= ''
     options.trial ?= false
-    "#{@apiServer}/oauth/authorize?client_id=#{@appId}&trial=#{options.trial}&response_type=code&scope=email&login_hint=#{options.loginHint}&redirect_uri=#{options.redirectURI}"
+    if options.state?
+      "#{@apiServer}/oauth/authorize?client_id=#{@appId}&trial=#{options.trial}&response_type=code&scope=email&login_hint=#{options.loginHint}&redirect_uri=#{options.redirectURI}&state=#{options.state}"
```
Comment (khamidou) [inline nylas.coffee]: I'd prefer not to duplicate the URL string. You should be doing something like this instead:

```coffeescript
url = "#{@apiServer}/oauth/authorize?client_id=#{@appId}&trial=#{options.trial}&response_type=code&scope=email&login_hint=#{options.loginHint}&redirect_uri=#{options.redirectURI}"

if options.state?
  url += "&state=#{options.state}"
```

**Comment by khamidou** [comment]
We're almost there! Fix the comment you've addressed and I'll merge this.

Thanks,

k

**Comment by slava-tre** [comment]
@khamidou Done...

**Comment by slava-tre** [comment]
Ok... @khamidou Could you confirm =)

**Comment by khamidou** [comment]
Thanks for fixing @slava-tre!

**Pull request merged**